community/kernel/src/main/java/org/neo4j/kernel/guard/TimeoutGuard.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.guard;
+
+import java.util.function.Supplier;
+
+import org.neo4j.helpers.Clock;
+import org.neo4j.kernel.impl.api.KernelStatement;
+import org.neo4j.kernel.impl.api.KernelTransactionImplementation;
+import org.neo4j.logging.Log;
+
+public class TimeoutGuard implements Guard
+{
+    private final Log log;
+    private Clock clock;
+
+    public TimeoutGuard( final Log log, Clock clock )
+    {
+        this.log = log;
+        this.clock = clock;
+    }
+
+    @Override
+    public void check( KernelStatement statement )
+    {
+        check( maxStatementCompletionTimeSupplier( statement ), "Statement timeout." );
+        check( statement.getTransaction() );
+    }
+
+    private void check (KernelTransactionImplementation transaction)
+    {
+        check( maxTransactionCompletionTimeSupplier( transaction ), "Transaction timeout." );
```

Comment:
Cleaned up. That was one of the things i left to measure later.
With lambda:
```TimeoutGuardBenchmark.checkStatement                               thrpt   25  108759.354 ±  3700.293  ops/ms
TimeoutGuardBenchmark.checkStatement:·gc.alloc.rate                thrpt   25    4977.462 ±   169.485  MB/sec```
without lambda (plain method calls):
```TimeoutGuardBenchmark.checkStatement                      thrpt   25  263592.702 ± 13171.769  ops/ms
TimeoutGuardBenchmark.checkStatement:·gc.alloc.rate       thrpt   25       0.001 ±     0.001  MB/sec```